npins/default.nix

```diff
@@ -26,21 +26,30 @@ let
       repository,
       revision,
       url ? null,
+      submodules,
       hash,
       branch ? null,
       ...
     }:
     assert repository ? type;
     # At the moment, either it is a plain git repository (which has an url), or it is a GitHub/GitLab repository
     # In the latter case, there we will always be an url to the tarball
-    if url != null then
-      (builtins.fetchTarball {
+    if url != null && !submodules then
+      builtins.fetchTarball {
         inherit url;
         sha256 = hash; # FIXME: check nix version & use SRI hashes
-      })
+      }
     else
-      assert repository.type == "Git";
       let
+        url =
+          if repository.type == "Git" then
```

Comment:
I can't really follow why we suddenlty need the switch on `repository.type` here. How does that relate to submodules and where did we have the "magic" GitHub URL format previously?

Reply:
So previously, a git repo was either some repo forge, which is guaranteed to have release URLs, or to be a generic git repo, in which case `fetchGit` would be used. Hence the assertion `assert repository.type == "Git";` previously in the code.

Now, due to submodules, we sometimes have to use `fetchGit` even for git forges, because the `url` would not have the desired content. So we have to reconstruct the full URL for the repo to use.